### DISCUSSION TITLE
Add animation to GradientBrush and/or GradientStop.

### DISCUSSION THREAD

**Discussion by Alshain01** [AvaloniaUI/Avalonia#12935] [Q&A]
**Is your feature request related to a problem? Please describe.**
Other UI frameworks, such as WPF, allow gradient stops to be animated.  As far as I can tell, Avalonia currently has no ability to animate a GradientBrush by manipulating the color and/or offset of a stop.

**Describe the solution you'd like**
Make the GradientBrush and/or GradientStop animatable.

**Describe alternatives you've considered**
The GradientBrush base class itself could be made animateable.  Instead of animating the offset property, the brush properties (such as the StartPoint and EndPoint for LinearGradientBrush) could be animated using transitions to achieve similar effects.  However, this may not be a reason to avoid adding it to the GradientStop itself.  It may be beneficial to add it to both as this would be more consistent with WPF, despite the differences in how animations work between the two.

**Additional context**
This would add many capabilities, however, gradient offset animation is frequently used to provide user feedback, particularly pointer over animations on buttons and other controls.
Consider the following potential use case which would produce a red button that sweeps a blue gradient diagonally across when the pointer moves over it:
```
<Button Width="100" Height="100">
  <Button.Background>
    <LinearGradientBrush StartPoint="0,0">
      <GradientStop Color="Red"/>
        <GradientStop Class="sweep" Color="Blue" Offset="0">
          <Transitions>
            <DoubleTransition Property="Offset" Duration="0:0:3"/>
          </Transitions>
        </GradientStop>
      <GradientStop Color="Red"/>
    </LinearGradientBrush>
  </Button.Background>
</Button>
```

```
<Style Selector="GradientStop.sweep:pointerover">
  <Setter Property="Offset" Value="1"/>
</Style>
```

**Comment by maxkatz6** [comment]
Moved to Q&A discussions.

> Other UI frameworks, such as WPF, allow gradient stops to be animated

Quite the opposite. In WPF you can't animate full gradients, only partial properties.
While in Avalonia, you can animate a whole gradient instead, including different gradient stop sets. Other properties are also interpolated as well.

So, you need to change your approach from trying to style a single property of the gradient stop, to using transitions of keystops to switch between sets of gradients:

https://github.com/AvaloniaUI/Avalonia/blob/master/samples/RenderDemo/Pages/AnimationsPage.xaml#L203-L229
https://github.com/AvaloniaUI/Avalonia/blob/master/samples/RenderDemo/Pages/TransitionsPage.xaml#L160-L176

Here's a bit old gifs of how it works https://github.com/AvaloniaUI/Avalonia/pull/6183

**Comment by Alshain01** [comment]
@maxkatz6 Ah I was so close.  Thanks for the links.  This was the actual effect I was looking for that I had working in WPF, but your post got me there, much appreciated.

```
<Style Selector="Border.Rect11">
  <Setter Property="Transitions">
    <Transitions>
      <BrushTransition Property="Background" Duration="0:0:0.3" />
    </Transitions>
  </Setter>
  <Setter Property="Background">
    <LinearGradientBrush StartPoint="0%,0%" EndPoint="0%,100%">
      <GradientStop Offset="0" Color="Red"/>
      <GradientStop Offset="1" Color="Blue"/>
      <GradientStop Offset="1" Color="Red"/>
    </LinearGradientBrush>
  </Setter>
</Style>

<Style Selector="Border.Rect11:pointerover">
  <Setter Property="Background" >
    <LinearGradientBrush StartPoint="0%,0%" EndPoint="0%,100%">
      <GradientStop Offset="0" Color="Red"/>
      <GradientStop Offset="0" Color="Blue"/>
      <GradientStop Offset="1" Color="Red"/>
    </LinearGradientBrush>
  </Setter>
</Style>
```

**Comment by Alshain01** [comment]
@maxkatz6 Well I spoke too soon.  This works on Borders but doesn't work at all on Buttons, I think because the transition is on the button but it's the background of the content presenter, but I'm not sure.  So I'm back to saying this isn't currently possible in Avalonia.  It does switch brushes, but no transition.

**Reply by timunie** [reply]
worth to prepare a minimal sample of what you expect to work and file it here. If we can reproduce, we may want to create an issue to track out of it.

**Reply by Alshain01** [reply]
There was an issue, it was closed.

Basically just take the xanl I posted and change it to a button instead of a border and it fails.

I've got it working now, it just took 10 times the code, a waste of a day, and a massive headache to do something that's so simple in WPF.  I ended up applying a data template to the content presenter of the button and basically recreating all the button styles within the button within the button and then hide the buttons native border with transparency.

**Reply by timunie** [reply]
What works for border should also work for buttons in general. I guess the issue is, that a /template/ child is beeing changed and not the actual background. 🤔 

https://github.com/AvaloniaUI/Avalonia/blob/cde91774ac9db803041642212eeb2626919c449d/src/Avalonia.Themes.Fluent/Controls/Button.xaml#L74-L78

**Reply by Alshain01** [reply]
Yes, sorry I meant to mention, that's exactly what it is, except the transition is on the button, not the content presenter, which is why the whole thing fails.  Changing the brush on the content presenter doesn't trigger the button's transition.   This is why I had to replace the whole visual.  Putting the transition on the content presenter didn't seem to work either.

**Comment by Alshain01** [comment]
Ok, I got it.  It was a nightmare but I had to create a Content template with the border and a custom type for it.  As much as I like Avalonia, this is definitely one area where WPF is superior.

**Comment by Coolkeke** [comment]
You may need this answer
```
<Border Height="40">
    <Border.Styles>
        <Style Selector="Border">
            <Style.Animations>
                <Animation
                    FillMode="None"
                    IterationCount="Infinite"
                    Duration="0:0:0.5">
                    <KeyFrame Cue="0%">
                        <Setter Property="Background">
                            <Setter.Value>
                                <LinearGradientBrush StartPoint="0%,100%" EndPoint="100%,100%">
                                    <GradientStop Offset="0" Color="#C2E4F6" />
                                    <GradientStop Offset="0" Color="#95CDEC" />
                                    <GradientStop Offset="1" Color="#C2E4F6" />
                                </LinearGradientBrush>
                            </Setter.Value>
                        </Setter>
                    </KeyFrame>
                    <KeyFrame Cue="100%">
                        <Setter Property="Background">
                            <Setter.Value>
                                <LinearGradientBrush StartPoint="0%,100%" EndPoint="100%,100%">
                                    <GradientStop Offset="0" Color="#C2E4F6" />
                                    <GradientStop Offset="1" Color="#95CDEC" />
                                    <GradientStop Offset="1" Color="#C2E4F6" />
                                </LinearGradientBrush>
                            </Setter.Value>
                        </Setter>
                    </KeyFrame>
                </Animation>
            </Style.Animations>
        </Style>
    </Border.Styles>
</Border>
